### DISCUSSION TITLE
Subnet-EVM Warp messages: message encoding in SendWarpMessage event logs

### DISCUSSION THREAD

**Discussion by Nuttymoon** [ava-labs/subnet-evm#794] [Q&A] [answered]
Hi, I am currently working on an experimental tool to monitor Avalanche Warp Messaging: [Avalanche Warp Navigator](https://github.com/AshAvalanche/avalanche-warp-navigator). The _navigator_ feature is added to the [Ash CLI](https://github.com/AshAvalanche/ash-rs/tree/main/crates/ash_cli) which is written in Rust.

To achieve this goal my first step was to decode the `data` of `SendWarpMessage` event logs to extract the `UnsignedMessage` and the `AddressedPayload`.

I managed to decode the messages and extract all the information:
![image](https://github.com/ava-labs/subnet-evm/assets/11584800/3344d73a-1449-416b-8638-cf356a13f75d)

Yet, I have some questions about the way the `UnsignedMessage` is encoded in the `SendWarpMessage` event `data`:
- What are the first 2 bytes of the `UnsignedMessage` used for? Is this for the codec version? (I saw a reference to `codecVersion` in the code)
- What are the 6 bytes after the length of the `AddressedPayload` used for?

Here is my current understanding of the encoding, which I validated by the messages I generated:
```python
'0000'     # 2B: CodecVersion?
'00003039' # 4B: NetworkID
'76dccb39c21a43aad4ffa98d4dd86a9ca29f5038a13b87658ef856bc161dbb47' # 32B: SourceChainID
# AddressesPayload
'00000062'     # 4B: AddressedPayload length
'000000000000' # 6B: ???
'8db97c7cece249c2b98bdc0226cc4c2a57bf52fc'                         # 20B: SourceAddress
'a4277d59ef210782b0b2a8043877a8bfcbc385942924c5547d3cd45493448b01' # 32B: DestinationChainID
'9632a79656af553f58738b0fb750320158495942'                         # 20B: DestinationAddress
'00000010466f722074686520456d7065726f7221'                         # XXB: Payload (ABI encoded)
```

Thanks!

**Comment by patrick-ogrady** [comment]
(side note: super cool project ❤️ )

**Comment by aaronbuchwald** [comment, marked as answer]
You can find the AWM message format here: https://github.com/ava-labs/avalanchego/blob/master/vms/platformvm/warp/README.md#awm-serialization and the Subnet-EVM payload serializations here: https://github.com/ava-labs/subnet-evm/blob/master/warp/payload/README.md.

**Reply by aaronbuchwald** [reply]
The message formats are still subject to change, so heads up that they are not set in stone yet.

**Comment by Nuttymoon** [comment]
Thanks for the links. I missed those READMEs somehow...
